### ISSUE TITLE
EntitySystem dependencies

### ISSUE THREAD

**Issue by gradientvera** [space-wizards/RobustToolbox#1822]
We need a way to specify (and automatically get) dependencies on EntitySystems.
Similar to [Dependency], and similar to [ComponentDependency] but not cursed

*I'm sorry Acruid, they managed to convince me*

**Comment by Acruid** [comment]
There are currently 2 ways to resolve EntitySystems, why do we need a 3rd way?

**Comment by gradientvera** [comment]
The idea is to cache these references to other entity systems so we don't have to go through the awful, slow static resolve methods, and also so we can explicitly see the dependencies between them.

Current `EntitySystem.Get<T>()` code:
```csharp
        public static T Get<T>() where T : IEntitySystem
        {
            return IoCManager.Resolve<IEntitySystemManager>().GetEntitySystem<T>();
        }
```

**Comment by Acruid** [comment]
There is no reason to be making IoC calls to resolve `IEntitySystemManager` when every `EntitySystem` already has a reference to it. The current 'proper' way to access another `EntitySystem` is with `EntitySystemManager.GetEntitySystem<T>()`. `IEntityManager` also has `IEntitySystemManager` as a public property, so `IEntityManager.EntitySysManager.GetEntitySystem<T>()` is how a component would access a system. You can get a `IEntityManager` reference basically anywhere.

I agree that those static helper functions on EntitySystem that go through IoC are terrible, and should be removed.

**Comment by PJB3005** [comment]
This would just be a convenience feature instead of having to manually resolve the entity systems with `IEntitySystemManager`. Good to me.

**Comment by Acruid** [comment]
Honestly a real DI pattern like constructor injection would be better than our wacky field reflection injection (for the same reasons the IoC field injection is bad). On the other hand, field reflection injection is way better than what it is currently doing, holding a ref to it's own container and resolving other objects inside of it at runtime. It is also really easy to refactor the field reflection to constructor injection if someone wants to fix it later.

The proposal isn't just convenience, this moves us in the right direction from a DI perspective, compared to the existing way. As far as I can tell there are no downsides to doing this. GO FOR IT!

**Comment by gradientvera** [comment]
Closed by #1890
